Andresmweber circleci/test automation feature

AndresMWeber opened the pull request:
So I'm not sure if you're interested in this functionality as it seems you have your own workflow for testing this and installing it worked out, however I updated your codebase to reflect proper setuptools/package structure.  Conforming to this standard enables proper pip/setuptools installation of the project to your maya directory or userScripts directory.

I also set it up so that it automates the testing for your project against Maya 2015-2018 using CircleCI and Docker images provided by [Marcus Ottosson](https://github.com/mottosso).  It currently is failing since the test suite doesn't pass all test cases: https://circleci.com/gh/AndresMWeber/node_calculator/7. 
 
I think you'll find it's a very robust workflow and now you can test your codebase without even opening maya in GUI mode and can run it straight from Maya's python interpreter.  I also left some boilerplate code in the Makefile for uploading your project to PyPi.

I've added some tips for installation/testing for any contributors in the README as well and removed the dependency on cmt, however you'll see in the Makefile that I've left cmt installed on the Docker image just in case you want to still use it.  I totally understand if this is a massive shift in your project and might cause you to reject the PR!

I'm not sure if this is what the current state of the project testing was, but just letting you know that as of right now the test suite fails (you'll see I also install/use nose as the test runner with the Makefile) with this report:

```
.FF....F....auto_unravel: True, auto_consolidate: True
auto_unravel: True, auto_consolidate: True
..11/01/2018 19:23:35 - node_calculator.logger - WARNING - Returned None for invalid node-property request of NcList instance: NcList([NcNode(A1, ['translateX']), 7.3, NcNode(B1, ['translateX'])]). Did you mean 'nodes'?
F...............E..EEEEEEE................................
======================================================================
ERROR: test_auto_consolidate (test_node_calculator.TestNodeCalculatorCore)
Test automatic attribute consolidating: .tx, .ty, .tz -> .t
----------------------------------------------------------------------
Traceback (most recent call last):
  File "D:\dev\node_calculator\tests\test_node_calculator.py", line 110, in test_auto_consolidate
    )[1]
TypeError: 'NoneType' object has no attribute '__getitem__'

======================================================================
ERROR: test_mdag_path (test_om_util.TestTracerClass)
Test mdag path creation & retrieving
----------------------------------------------------------------------
Traceback (most recent call last):
  File "D:\dev\node_calculator\tests\test_om_util.py", line 99, in setUp
    self.test_instance = cmds.instance(self.test_node, name=self.instance_name)[0]
ValueError: No object matches name: |testNode

======================================================================
ERROR: test_mobj (test_om_util.TestTracerClass)
Test mobj creation & retrieving
----------------------------------------------------------------------
Traceback (most recent call last):
  File "D:\dev\node_calculator\tests\test_om_util.py", line 99, in setUp
    self.test_instance = cmds.instance(self.test_node, name=self.instance_name)[0]
ValueError: No object matches name: |testNode

======================================================================
ERROR: test_node (test_om_util.TestTracerClass)
Test node creation & retrieving
----------------------------------------------------------------------
Traceback (most recent call last):
  File "D:\dev\node_calculator\tests\test_om_util.py", line 99, in setUp
    self.test_instance = cmds.instance(self.test_node, name=self.instance_name)[0]
ValueError: No object matches name: |testNode

======================================================================
ERROR: test_plug (test_om_util.TestTracerClass)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "D:\dev\node_calculator\tests\test_om_util.py", line 99, in setUp
    self.test_instance = cmds.instance(self.test_node, name=self.instance_name)[0]
ValueError: No object matches name: |testNode

======================================================================
ERROR: test_shape (test_om_util.TestTracerClass)
Test shape creation & retrieving
----------------------------------------------------------------------
Traceback (most recent call last):
  File "D:\dev\node_calculator\tests\test_om_util.py", line 99, in setUp
    self.test_instance = cmds.instance(self.test_node, name=self.instance_name)[0]
ValueError: No object matches name: |testNode

======================================================================
ERROR: test_speed (test_om_util.TestTracerClass)
Test speed advantage of om_util functions over cmds equivalent
----------------------------------------------------------------------
Traceback (most recent call last):
  File "D:\dev\node_calculator\tests\test_om_util.py", line 99, in setUp
    self.test_instance = cmds.instance(self.test_node, name=self.instance_name)[0]
ValueError: No object matches name: |testNode

======================================================================
ERROR: test_split_plug_string (test_om_util.TestTracerClass)
Test plug strings are correctly split into their components
----------------------------------------------------------------------
Traceback (most recent call last):
  File "D:\dev\node_calculator\tests\test_om_util.py", line 99, in setUp
    self.test_instance = cmds.instance(self.test_node, name=self.instance_name)[0]
ValueError: No object matches name: |testNode

======================================================================
FAIL: test_transform_creation (test_convenience.TestCreateNode)
Test node creation via noca-convenience function
----------------------------------------------------------------------
Traceback (most recent call last):
  File "D:\dev\node_calculator\tests\test_convenience.py", line 140, in test_transform_creation
    self.assertEqual(node.node, name)
AssertionError: u'test1' != 'test'

======================================================================
FAIL: test_creation (test_convenience.TestLocator)
Test locator creation via noca-convenience function
----------------------------------------------------------------------
Traceback (most recent call last):
  File "D:\dev\node_calculator\tests\test_convenience.py", line 95, in test_creation
    self.assertEqual(node.node, name)
AssertionError: u'test2' != 'test'

======================================================================
FAIL: test_creation (test_convenience.TestTransform)
Test transform creation via noca-convenience function
----------------------------------------------------------------------
Traceback (most recent call last):
  File "D:\dev\node_calculator\tests\test_convenience.py", line 119, in test_creation
    self.assertEqual(node.node, name)
AssertionError: u'test3' != 'test'

======================================================================
FAIL: test_noca_methods (test_nc_list.TestNcListClass)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "D:\dev\node_calculator\tests\test_nc_list.py", line 117, in test_noca_methods
    self.assertListEqual(self.nc_list.nodes, TEST_NODES[:2])
AssertionError: Lists differ: [u'A1', u'B1'] != ['A', 'B']

First differing element 0:
A1
A

- [u'A1', u'B1']
?  -  -   -  -

+ ['A', 'B']

----------------------------------------------------------------------
Ran 72 tests in 0.287s

FAILED (failures=4, errors=8)
```

AndresMWeber commented:
Ah you're right.  I just skimmed the fails and didn't actually run it within Maya but you're totally right, persisting scene objects are causing duplicate name + '1' issues.  

I made a context manager to delete each test's created objects in my own repo's test cases so that is definitely the culprit.  Let me look into it and see if I can't get it to pass given that new information!

mischakolbe commented:
> Ah you're right. I just skimmed the fails and didn't actually run it within Maya but you're totally right, persisting scene objects are causing duplicate name + '#' issues.
> 
> I made a context manager to delete each test's created objects in my own repo's test cases so that is definitely the culprit. Let me look into it and see if I can't get it to pass given that new information!

There's a really good chance that a simple tearDown in the erroring tests would fix most/all issues. I guess CMT didn't require me to work clean, so it ended up sloppy :S :D

mischakolbe commented on tests/base.py:
```diff
@@ -0,0 +1,11 @@
+from unittest import TestCase
+
+
+class BaseTestCase(TestCase):
```
Good idea to create this class in the first place. Might come in handy for something other than the setUp/tearDown stuff as well. Thanks!!

AndresMWeber replied:
Yeah I find that having a base TestCase class is almost always handy for projects!  Marking this as resolved.

mischakolbe commented:
> 
> 
> Just added a new `BaseTestCase` class with implemented **tearDown** and just in case a **setUpClass** that force create new maya scenes in this [commit](https://github.com/mischakolbe/node_calculator/pull/66/commits/dd476d94ba9d1546183ea1e3ddd5f284bff933b4)
> 
> Here's the status of the [**build on CircleCI**](https://circleci.com/gh/AndresMWeber/node_calculator/8).
> Great news: now all tests pass, however ONE is still failing, maybe you have some insight into (I'd investigate, however I don't have enough spare time at the moment):
> 
> ```
> ................E.........................................
> ======================================================================
> ERROR: test_auto_consolidate (tests.test_node_calculator.TestNodeCalculatorCore)
> Test automatic attribute consolidating: .tx, .ty, .tz -> .t
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "D:\dev\node_calculator\tests\test_node_calculator.py", line 110, in test_auto_consolidate
>     )[1]
> TypeError: 'NoneType' object has no attribute '__getitem__'
> 
> ----------------------------------------------------------------------
> Ran 72 tests in 0.624s
> 
> FAILED (errors=1)
> ```

Not sure what's going on here, to be honest. I'll have to take a look and debug this, hopefully tomorrow.

mischakolbe commented on tests/base.py:
```diff
@@ -0,0 +1,11 @@
+
+class BaseTestCase(TestCase):
+    @classmethod
+    def setUpClass(self):
```
I believe this should be `setUp` instead of `setUpClass`. The way I wrote a couple of the tests it should be safer to reset before every method, rather than every class.

Feel free to leave this as is though. There would be the need to add a `super(XYZ, self).setUp()` to all the classes with their setUp. I could take part of these annoying tasks :)

AndresMWeber replied:
I used setUpClass so it would just make sure that the first test run in the TestCase suite would be working from a new scene, just in case whatever TestCase that was running before it didn't clean up after itself.  I figured it would be overkill to do both setup and teardowns for new scene creation.  I can update it to setUp instead if you'd like though!

mischakolbe commented on tests/base.py:
```diff
@@ -0,0 +1,11 @@
+import maya.cmds as mc
```
To keep the `cmds` vs `mc` consistent it would be great to use `cmds` here, too. Otherwise it's the one exception :)

AndresMWeber replied:
Addressed this in the most recent commit.  Marking as resolved.

AndresMWeber commented:
Just fixed the test...I didn't notice you had defined a tearDown method within that TestCase.  Just had to add a call to the super's TearDown.  Also fixed the mc->cmds.

AndresMWeber commented:
Just added a commit that addresses mc->cmds and fixed the error (just had to call super tearDown since I didn't see you had defined a custom tearDown within that TestCase).

[Passing build on CircleCI!
](https://circleci.com/gh/AndresMWeber/node_calculator/12)
[![CircleCI](https://circleci.com/gh/AndresMWeber/node_calculator/tree/andresmweber-circleci%2Ftest-automation-feature.svg?style=svg)](https://circleci.com/gh/AndresMWeber/node_calculator/tree/andresmweber-circleci%2Ftest-automation-feature)

AndresMWeber commented:
I completely get it, if you have any questions don't hesitate to reach out.  I know exactly what you mean, production life makes for very little github time!  I hope I'll get a chance to contribute again, but until you roll out your major changes for v3 I will keep an eye out.

mischakolbe commented:
Ciao @AndresMWeber,

I managed to address some other open issues this weekend. I did that on top of your additions. Potentially I could just merge this new branch, which includes your changes:
https://github.com/mischakolbe/node_calculator/tree/weekend_fixeroo_1

Let me know how you'd prefer to go about this :)

mischakolbe commented:
Sorry, now I'm spamming you, @AndresMWeber.
Did some of the tests error for you, too when you ran them locally via mayapy?

I'm getting some strange errors:

```
======================================================================
ERROR: test_convenience (unittest.loader.ModuleImportFailure)
----------------------------------------------------------------------
ImportError: Failed to import test module: test_convenience
Traceback (most recent call last):
  File "C:\Program Files\Autodesk\Maya2018\bin\python27.zip\unittest\loader.py", line 254, in _find_tests
    module = self._get_module_from_name(name)
  File "C:\Program Files\Autodesk\Maya2018\bin\python27.zip\unittest\loader.py", line 232, in _get_module_from_name
    __import__(name)
  File "E:\Dropbox\__SoftwareSpecific__\Maya\scripts\node_calculator\tests\test_convenience.py", line 15, in <module>
    import node_calculator.core as noca
ImportError: No module named node_calculator.core


======================================================================
ERROR: test_nc_attrs (unittest.loader.ModuleImportFailure)
----------------------------------------------------------------------
ImportError: Failed to import test module: test_nc_attrs
Traceback (most recent call last):
  File "C:\Program Files\Autodesk\Maya2018\bin\python27.zip\unittest\loader.py", line 254, in _find_tests
    module = self._get_module_from_name(name)
  File "C:\Program Files\Autodesk\Maya2018\bin\python27.zip\unittest\loader.py", line 232, in _get_module_from_name
    __import__(name)
  File "E:\Dropbox\__SoftwareSpecific__\Maya\scripts\node_calculator\tests\test_nc_attrs.py", line 15, in <module>
    import node_calculator.core as noca
ImportError: No module named node_calculator.core
```

However; I use `import node_calculator.core as noca` in all other tests, too and they work fine. It's as if some tests start running early, before Maya is properly loaded.

AndresMWeber commented:
Sounds perfect to me and sorry for being inactive for so long.  If you want to reject this PR and just merge your other branch I think that makes a lot of sense.  What a quirky Maya-specific issue the test suite local run ran into...  I can't say I can reproduce it, maybe it had to do with what directory you ran the tests from?  But that shouldn't really make sense if other tests run fine...  

I'm pretty sure pymel has some kind of blocker set up so that it waits until maya is fully loaded so possibly replacing the maya import with a pymel import in the base test will block the test suite artificially until maya is loaded and remove the bug and the necessity to add the path to sys.path.  Just a thought!

mischakolbe commented:
Haha, in comparison you responded instantly ;P

Ok, I'll merge the other branch then. Got 1-2 other things to address, so I'll do all that in one go.

Could totally be from where I ran it from or so... But yeah; the fact that it ran most tests, except for 2 that are "identical" in their setup is super strange.

I thought about trying to add some blocker until it's fully loaded, but I haven't done anything alike so far. The PyMel is a cheeky trick, I might try that :D

AndresMWeber commented:
@mischakolbe also here's a really good synopsis of CI that might be helpful https://realpython.com/python-continuous-integration/

mischakolbe commented:
Thanks a lot, @AndresMWeber ! I'll definitely check it out more in-depth asap :P

Just merged the PR and about to update the tutorial videos.
Hopefully I'll get to the other issues tonight as well.

Quick question; is this by any chance your Twitter account?
https://twitter.com/andresweber
If so I'll mention you when I do the notification about the update.

AndresMWeber commented:
@mischakolbe Yeah that's mine, I don't really use it but feel free to add it to the notification.

mischakolbe commented:
That's what I thought, too in the end, @AndresMWeber :D
I linked to your gitHub instead, that probably makes more sense :)
Thanks again for your PR! So awesome to have others contribute to the project!
